From 0c82ce44cfb74c07d9e23b4f528ad1f7798e43bb Mon Sep 17 00:00:00 2001
From: naoto <nnomura98@gmail.com>
Date: Tue, 11 Oct 2022 14:48:55 +0200
Subject: [PATCH 07/10] password top offset

Added option to add offset from top to chars in password mode (to make centering chars possible).
---
 dmenu.1 | 4 ++++
 dmenu.c | 5 ++++-
 2 files changed, 8 insertions(+), 1 deletion(-)

diff --git a/dmenu.1 b/dmenu.1
index fcf72c2..17533ad 100644
--- a/dmenu.1
+++ b/dmenu.1
@@ -31,6 +31,8 @@ dmenu \- dynamic menu
 .IR color ]
 .RB [ \-w
 .IR windowid ]
+.RB [ \-Po
+.IR offset ]
 .RB [ \-d
 .IR separator ]
 .RB [ \-D
@@ -64,6 +66,8 @@ dmenu displays matching lines below prompt
 .TP
 .B \-P
 dmenu will not directly display the keyboard input, but instead replace it with dots. All data from stdin will be ignored.
+.BI \-Po " offset"
+password top offset
 .TP
 .BI \-l " lines"
 dmenu lists items vertically, with the given number of lines.
diff --git a/dmenu.c b/dmenu.c
index e45ca09..ca46812 100644
--- a/dmenu.c
+++ b/dmenu.c
@@ -53,6 +53,7 @@ static struct item *matches, *matchend;
 static struct item *prev, *curr, *next, *sel;
 static int mon = -1, screen;
 static int lines_below_prompt = 0;
+static int password_top_offset = 0;
 
 static Atom clip, utf8;
 static Display *dpy;
@@ -173,7 +174,7 @@ drawmenu(void)
 	if (passwd) {
 		censort = ecalloc(1, sizeof(text));
 		memset(censort, '*', strlen(text));
-		drw_text(drw, x, 0, w, bh, lrpad / 2, censort, 0);
+		drw_text(drw, x, password_top_offset*fh/100, w, bh, lrpad / 2, censort, 0);
 		curpos = TEXTW(censort) - TEXTW(&censort[cursor]);
 		free(censort);
 	} else {
@@ -899,6 +900,8 @@ main(int argc, char *argv[])
 			colors[SchemeSel][ColFg] = argv[++i];
 		else if (!strcmp(argv[i], "-w"))   /* embedding window id */
 			embed = argv[++i];
+		else if (!strcmp(argv[i], "-Po"))   /* password top offset */
+			password_top_offset = atoi(argv[++i]);
 		else if (!strcmp(argv[i], "-d") || /* field separator */
 		         !strcmp(argv[i], "-D")) {
 			sepchr = argv[i][1] == 'D' ? strrchr : strchr;
-- 
2.38.0

